renku/api/datasets.py

```diff
@@ -352,69 +356,103 @@ def _add_from_git(self, dataset, path, url, target, **kwargs):
-            files = []
-            dst.mkdir(parents=True, exist_ok=True)
-            # FIXME get all files from submodule index
-            for f in src.iterdir():
-                try:
-                    files.extend(
-                        self._add_from_git(
-                            dataset,
-                            path,
-                            url,
-                            target=f.relative_to(submodule_path),
-                            **kwargs
-                        )
-                    )
-                except ValueError:
-                    pass  # skip files outside the relative path
-            return files
+            files.extend({e.path for e in src_root.traverse()})
+
+        # create a map of all submodule's dataset files to be used to
+        # get creators. Use Dataset's creator if a file does not have
+        # a creator
+        submodule_client = LocalClient(submodule_path)
+        submodule_dataset_creators = {}
+        for sb_dataset in submodule_client.datasets.values():
+            for f in sb_dataset.files:
+                creators = f.creator if f.creator else sb_dataset.creator
+                submodule_dataset_creators[str(f.path)] = creators
+
+        results = []
+
+        for entry in files:
+            src = submodule_path / entry
+            dst_target = Path(entry).relative_to(relative_to)
+            # link the target into the data directory
+            dst = self.path / dataset_path / dst_target
+
+            # Ignore '.gitmodules'
+            if os.path.basename(entry) == '.gitmodules':
+                continue
 
-        if not dst.parent.exists():
-            dst.parent.mkdir(parents=True)
+            if src.is_dir():
+                dst.mkdir(parents=True, exist_ok=True)
+            else:
+                dst.parent.mkdir(parents=True, exist_ok=True)
 
-        os.symlink(os.path.relpath(str(src), str(dst.parent)), str(dst))
+                os.symlink(
+                    os.path.relpath(str(src), str(dst.parent)), str(dst)
+                )
 
-        # grab all the creators from the commit history
-        git_repo = Repo(str(submodule_path.absolute()))
-        creators = []
-        for commit in git_repo.iter_commits(paths=target):
-            creator = Creator.from_commit(commit)
-            if creator not in creators:
-                creators.append(creator)
+                creators = []
+                # use creator from the original dataset if it exists
+                file_creator = submodule_dataset_creators.get(entry)
+                if file_creator:
+                    creators = file_creator
+                else:
+                    # grab all the creators from the commit history
+                    for commit in submodule_repo.iter_commits(
+                        paths=dst_target
+                    ):
+                        creator = Creator.from_commit(commit)
+                        if creator not in creators:
```

Comment:
if you use a `set()` you don't need to do this check.

Reply:
I prefer to keep it as it is because then I have to convert from set to dict.